---
title: "botero_analysis"
author: "Sean Lu"
date: "`r Sys.Date()`"
output: pdf_document
editor_options:
  chunk_output_type: console
---
```{r knitr options, include = FALSE}
knitr::opts_chunk$set(cache=FALSE, cache.path=".cache/")
```

## Setup
```{r Loading in Libraries}
library("phyloseq")
library("ggplot2")     
library("dplyr")        
library("tibble")
library("ggpubr")
library("phylosmith")
library("DESeq2")
library("EnhancedVolcano")
library("microbiome")
library("eulerr")
library("ggVennDiagram")
```


## Load in processed Data
```{r, eval=TRUE, include=TRUE}
ps_dada2 <- readRDS("data_processed/botero_2014/ps_dada2.rds")
ps_qiime2 <- readRDS("data_processed/botero_2014/ps_qiime2.rds")
ps_ms <- readRDS("data_processed/botero_2014/ps_metascope_priors_trimmed.rds")

ps_dada2_oral <- subset_samples(ps_dada2, Sample_type == "Oropharynx")
ps_qiime2_oral <- subset_samples(ps_qiime2, Sample_type == "Oropharynx")
ps_ms_oral <- subset_samples(ps_ms, Sample_type == "Oropharynx")
ps_dada2_nasal <- subset_samples(ps_dada2, Sample_type == "Nasal")
ps_qiime2_nasal <- subset_samples(ps_qiime2, Sample_type == "Nasal")
ps_ms_nasal <- subset_samples(ps_ms, Sample_type == "Nasal")

# Use ps_melt to generate tidy format dataframes
dada2_df <- psmelt(ps_dada2) |> 
  dplyr::mutate(Species = ifelse(is.na(Species), NA, paste0(Genus, " ", Species))) |>
  dplyr::rename(Superkingdom = Kingdom)
dada2_df$pipeline = "DADA2"

qiime2_df <- psmelt(ps_qiime2) |>
  dplyr::select(-Confidence)
qiime2_df$pipeline = "QIIME2"

ms_df <- psmelt(ps_ms)
ms_df$pipeline = "MetaScope"
ms_df$Superkingdom = "Bacteria"
ms_df <- ms_df |>
  dplyr::relocate(Superkingdom, .before = phylum)
colnames(ms_df) <- c("OTU", "Sample", "Abundance", "Sequencing_Type", "Patient",
                     "Sample_type", "status", "Superkingdom", "Phylum", "Class", 
                     "Order", "Family", "Genus", "Species", "pipeline")
# Merge all data together
merged_df <- rbind(dada2_df, qiime2_df, ms_df)


# Generate relative abundance psmelt dataframes
dada2_relab_df <- phylosmith::relative_abundance(ps_dada2) |>
  phyloseq::psmelt() |> 
  dplyr::mutate(Species = ifelse(is.na(Species), NA, paste0(Genus, " ", Species))) |>
  dplyr::rename(Superkingdom = Kingdom)
dada2_relab_df$pipeline = "DADA2"

qiime2_relab_df <- phylosmith::relative_abundance(ps_qiime2) |>
  phyloseq::psmelt() |>
  dplyr::select(-Confidence)
qiime2_relab_df$pipeline = "QIIME2"

ms_relab_df <- phylosmith::relative_abundance(ps_ms) |>
  phyloseq::psmelt()
ms_relab_df$pipeline = "MetaScope"
ms_relab_df$Superkingdom = "Bacteria"
ms_relab_df <- ms_relab_df |>
  dplyr::relocate(Superkingdom, .before = phylum)
colnames(ms_relab_df) <- c("OTU", "Sample", "Abundance", "Sequencing_Type", "Patient",
                     "Sample_type", "status", "Superkingdom", "Phylum", "Class", 
                     "Order", "Family", "Genus", "Species", "pipeline")

merged_relab_df <- rbind(dada2_relab_df, qiime2_relab_df, ms_relab_df)
```
The DADA2 data is generated from the `dada2_botero.Rmd` file. The MetaScope data 
was generated from the `process_metascope_id.R` functions. 

## Plotting relative abundances of MetaScope and DADA2
### Phylum Level Abundances
```{r, Phylum level Abundances}
top_phyla <- merged_relab_df |>
  dplyr::group_by(Phylum) |>
  dplyr::summarise(total_abund = sum(Abundance), .groups = "drop") |>
  dplyr::filter(!is.na(Phylum)) |>
  slice_max(order_by=total_abund, n = 5) |>
  pull(Phylum) |>
  sort()

relab_phylum <- ggplot(merged_df, 
                        aes(x = Sample, y = Abundance, fill = Phylum)) + 
  geom_bar(position ="fill",stat = "identity") + 
  facet_grid(cols = vars(Sample_type, status, pipeline), scales = "free_x") +
  scale_fill_discrete() + 
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
relab_phylum
relab_phylum_legend <- get_legend(relab_phylum)

merged_relab_df |>
  dplyr::filter(Phylum %in% top_phyla) |>
  ggplot(aes(fill=status, y=Abundance, x=Phylum)) + 
  geom_boxplot() + 
  facet_grid(vars(Sample_type, pipeline)) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  scale_y_continuous(trans='log2')
```

```{r Saving stacked phylum relab figures, eval=FALSE, include=FALSE}
ggsave(file="figures/p3_stacked_relab_phylum.svg", 
       plot=relab_genus + theme(legend.position = "none",
           axis.text.x=element_blank(),
           axis.ticks.x=element_blank()),
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)
ggsave(file="figures/p3_legend.svg", plot=as_ggplot(relab_phylum_legend),
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)
```
At Phylum level taxonomies, both DADA2 and MetaScope show similar results. Nasal 
samples in the controls show increased abundance of Acidobacteriota compared to 
TB sample and a decreased relative abundance in Pseudomonadota in controls 
relative to TB samples. The oropharynx samples mild decrease in Baciollota phyla
and increases in Fusobacteriota and Pseudomonadota in the controls compared to 
TB positive samples. 

### Genus Level Abundances
```{r, Genus level Abundances}
top_genera <- merged_relab_df |>
  dplyr::group_by(Genus) |>
  dplyr::summarise(total_abund = sum(Abundance), .groups = "drop") |>
  dplyr::filter(!is.na(Genus)) |>
  slice_max(order_by=total_abund, n = 25) |>
  pull(Genus) |>
  sort()

relab_genus <- ggplot(merged_df, 
                        aes(x = Sample, y = Abundance, fill = Genus)) + 
  geom_bar(position ="fill",stat = "identity") + 
  facet_grid(cols = vars(Sample_type, status, pipeline), scales = "free_x") +
  scale_fill_discrete(breaks = top_genera) + 
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
relab_genus
relab_genus_legend <- get_legend(relab_genus)

top_genera <- merged_relab_df |>
  dplyr::group_by(Genus) |>
  dplyr::summarise(total_abund = sum(Abundance), .groups = "drop") |>
  dplyr::filter(!is.na(Genus)) |>
  slice_max(order_by=total_abund, n = 5) |>
  pull(Genus) |>
  sort()

merged_relab_df |>
  dplyr::filter(Genus %in% top_genera) |>
  ggplot(aes(fill=status, y=Abundance, x=Genus)) + 
  geom_boxplot() +
  facet_grid(vars(Sample_type, pipeline)) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  scale_y_continuous(trans='log2')
```

```{r Saving stacked genus relab figures, eval=FALSE, include=FALSE}
ggsave(file="figures/p2_stacked_relab_genus.svg", 
       plot=relab_genus,
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)
ggsave(file="figures/p2_legend.svg", plot=as_ggplot(relab_genus_legend),
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)
```
At the genus level, still DADA2 and MetaScope have similar relative abundances
and identify the same genera that are differentially expressed. Notably, DADA2 
identifies more Actinobacillus 

### Species Level Abundances
```{r Species Level Abundances}
top_species <- merged_relab_df |>
  dplyr::group_by(Species, Genus) |>
  dplyr::summarise(total_abund = sum(Abundance), .groups = "drop") |>
  dplyr::filter(!is.na(Species)) |>
  slice_max(order_by=total_abund, n = 25) |>
  pull(Species) |>
  sort()

relab_species <- ggplot(merged_df, 
                        aes(x = Sample, y = Abundance, fill = Species)) + 
  geom_bar(position ="fill",stat = "identity") + 
  facet_grid(cols = vars(Sample_type, status, pipeline), scales = "free_x") +
  scale_fill_discrete(breaks = top_species) + 
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
relab_species
relab_species_legend <- get_legend(relab_species)

merged_relab_df |>
  dplyr::filter(Species %in% top_species) |>
  ggplot(aes(fill=status, y=Abundance, x=Species)) + 
  geom_boxplot() + 
  facet_grid(vars(Sample_type, pipeline)) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  scale_y_continuous(trans='log2')
```

```{r Saving stacked species relab figures, eval=FALSE, include=FALSE}
ggsave(file="figures/p1_stacked_relab.svg", 
       plot=relab_species,
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)

ggsave(file="figures/p1_legend.svg", plot=as_ggplot(relab_species_legend),
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)
```



## Filtered abundance barplots
```{r Filtered abundance plots}
high_abund_dada2 <- merged_df |> 
  dplyr::filter(Abundance > 1000) |>
  dplyr::filter(pipeline == "DADA2")

high_abund_ms <- merged_df |> 
  dplyr::filter(Abundance > 1000) |>
  dplyr::filter(pipeline == "MetaScope")

select_species_both <- unique(high_abund_dada2$Species[high_abund_dada2$Species %in% high_abund_ms$Species])
select_species_dada2 <- unique(high_abund_dada2$Species[!(high_abund_dada2$Species %in% high_abund_ms$Species)])
select_species_ms <- unique(high_abund_ms$Species[!(high_abund_ms$Species %in% high_abund_dada2$Species)])

ms_unique_species <- merged_df |> 
  dplyr::filter(Species %in% select_species_ms) |>
  ggplot(aes(fill=Sample_type, y=Abundance, x=Species)) + 
  geom_bar(position="dodge", stat="identity") + 
  facet_grid(vars(pipeline), vars(status)) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
ms_unique_species
  
dada2_unique_species <- merged_df |> 
  dplyr::filter(Species %in% select_species_dada2) |>
  ggplot(aes(fill=Sample_type, y=Abundance, x=Species)) + 
  geom_bar(position="dodge", stat="identity") + 
  facet_grid(vars(pipeline), vars(status)) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
dada2_unique_species

high_abund_species <- merged_df |> 
  dplyr::filter(Species %in% select_species_both) |>
  ggplot(aes(fill=Sample_type, y=Abundance, x=Species)) + 
  geom_bar(position="dodge", stat="identity") + 
  facet_grid(vars(pipeline), vars(status)) + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
high_abund_species
```

```{r Saving high abundance plots, eval=FALSE, include=FALSE}
ggsave(file="figures/p4_ms_unique_species.svg", plot=ms_unique_species,
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)

ggsave(file="figures/p5_dada2_unique_species.svg", plot=dada2_unique_species,
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)

ggsave(file="figures/p6_high_abund_species", plot=high_abund_species,
       device = "svg", width=7.5, height=4, units="in",
       dpi = 300)
```

## Heatmaps
```{r Filtered heatmaps}
ps_ms_filt <- taxa_filter(ps_ms, frequency = 0.05)
  
abundance_heatmap(ps_ms_filt, classification = 'genus',
  treatment = "Sample_type", transformation = 'log10') + 
  facet_wrap(vars(status,Sample_type), nrow = 1, scales = "free_x")

ps_dada2_filt <- taxa_filter(ps_dada2, frequency = 0.01)
abundance_heatmap(ps_dada2_filt, classification = 'Genus',
  treatment = "Sample_type", transformation = 'log2') + 
  facet_wrap(vars(status,Sample_type), nrow = 1, scales = "free_x")
```

## Plotting Alpha Diversity
```{r Alpha Diversity}
p2_1 <- plot_richness(ps_dada2_oral, measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"), x = "status") + 
  stat_compare_means(label = "p.signif", label.x = 1.5, comparisons = list(c("Control", "TB")),
                     method = "wilcox.test") + 
  geom_boxplot()

p2_2 <- plot_richness(ps_ms_oral, measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"), x = "status")+ 
  stat_compare_means(label = "p.signif", label.x = 1.5, comparisons = list(c("Control", "TB")),
                     method = "wilcox.test") + 
  geom_boxplot()

p2_3 <- plot_richness(ps_dada2_nasal, measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"), x = "status") + 
  stat_compare_means(label = "p.signif", label.x = 1.5, comparisons = list(c("Control", "TB")), method = "wilcox.test") + 
  geom_boxplot()

p2_4 <- plot_richness(ps_ms_nasal, measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"), x = "status") + 
  stat_compare_means(label = "p.signif", label.x = 1.5, comparisons = list(c("Control", "TB")),method = "wilcox.test") + 
  geom_boxplot()

ggarrange(p2_1, p2_2, p2_3, p2_4, labels = "AUTO")
```

## PCOA Plots
```{r PCoA plots}
ps_ms.ord <- ordinate(ps_ms, "PCoA", "bray")
p3_1 <- plot_ordination(ps_ms, ps_ms.ord, type="samples", color="status", shape="Sample_type", title="MetaScope Analysis") + 
  geom_point(size=3) +
  stat_ellipse(aes(linetype=Sample_type)) 
  
set.seed(123)
ps_dada2.ord <- ordinate(ps_dada2, "PCoA", "bray")
p3_2 <- plot_ordination(ps_dada2, ps_dada2.ord, type="samples", color="status", shape="Sample_type", title="DADA2 Analysis") + 
  geom_point(size=3) + 
  stat_ellipse(aes(linetype=Sample_type))

set.seed(124)
ps_qiime2.ord <- ordinate(ps_qiime2, "PCoA", "bray")
p3_3 <- plot_ordination(ps_qiime2, ps_qiime2.ord, type="samples", color="status", shape="Sample_type", title="QIIME2 Analysis") + 
  geom_point(size=3) + 
  stat_ellipse(aes(linetype=Sample_type))
ggarrange(p3_1, p3_2, p3_3, labels = "AUTO", common.legend = TRUE, legend = "bottom")

```

## Core Microbiome Analysis
```{r}
ps_dada2_rel <- microbiome::transform(ps_dada2, "compositional")
ps_ms_rel <- microbiome::transform(ps_ms, "compositional")

list_core_ms <- c()

groups <- list(c("TB", "Sputum"), c("TB", "Nasal"), c("TB", "Oropharynx"),
               c("Control", "Nasal"), c("Control", "Oropharynx"))
for (i in 1:5){
  ps.sub <- subset_samples(ps_ms_rel, status == groups[[i]][1] & Sample_type == groups[[i]][2])
  core_m <- core_members(ps.sub,
                         detection = 0.001, 
                         prevalence = 0.2)
  list_core_ms[[i]] <- core_m
}
names(list_core_ms) <- lapply(groups, function(x) paste(x, collapse = " "))
p4_1 <- ggVennDiagram(list_core_ms, label_geom = "text", label = "count") + 
  scale_fill_distiller(palette = "Blues") + 
  scale_x_continuous(expand = expansion(mult = .2)) + 
  ggtitle("MetaScope Core Microbiome")

list_core_dada <- c() # an empty object to store information
for (i in 1:5){
  ps.sub <- subset_samples(ps_dada2_rel, status == groups[[i]][1] & Sample_type == groups[[i]][2])
  core_m <- core_members(ps.sub,
                         detection = 0.001,
                         prevalence = 0.2)
  list_core_dada[[i]] <- core_m
}
names(list_core_dada) <- lapply(groups, function(x) paste(x, collapse = " "))
p4_2 <- ggVennDiagram(list_core_dada, label_geom = "text", label = "count") + 
  scale_fill_distiller(palette = "Blues") + 
  scale_x_continuous(expand = expansion(mult = .2)) + 
  ggtitle("DADA2 Core Microbiome")

ggarrange(p4_1, p4_2, labels = "AUTO", common.legend = TRUE, legend = "right")
```

```{r}
deseq_ms_oral = phyloseq_to_deseq2(ps_ms_oral, ~ status)
deseq_ms_oral = estimateSizeFactors(deseq_ms_oral, type = 'poscounts')
deseq_ms_oral = DESeq(deseq_ms_oral, test="Wald", fitType="parametric")
deseq_ms_oral_res = results(deseq_ms_oral, cooksCutoff = FALSE)
deseq_ms_oral_res <- cbind(deseq_ms_oral_res, 
                           as(tax_table(ps_ms_oral)[rownames(deseq_ms_oral_res), ], "matrix"))

EnhancedVolcano(deseq_ms_oral_res,
  lab = deseq_ms_oral_res$species,
  x = 'log2FoldChange',
  y = 'pvalue')

deseq_ms_nasal = phyloseq_to_deseq2(ps_ms_nasal, ~ status)
deseq_ms_nasal = estimateSizeFactors(deseq_ms_nasal, type = 'poscounts')
deseq_ms_nasal = DESeq(deseq_ms_nasal, test="Wald", fitType="parametric")
deseq_ms_nasal_res = results(deseq_ms_nasal, cooksCutoff = FALSE)
deseq_ms_nasal_res <- cbind(deseq_ms_nasal_res, 
                           as(tax_table(ps_ms_nasal)[rownames(deseq_ms_nasal_res), ], "matrix"))

EnhancedVolcano(deseq_ms_nasal_res,
  lab = deseq_ms_nasal_res$species,
  x = 'log2FoldChange',
  y = 'pvalue')


deseq_dada2_oral = phyloseq_to_deseq2(ps_dada2_oral, ~ status)
deseq_dada2_oral = estimateSizeFactors(deseq_dada2_oral, type = 'poscounts')
deseq_dada2_oral = DESeq(deseq_dada2_oral, test="Wald", fitType="parametric")

deseq_dada2_oral_res = results(deseq_dada2_oral, cooksCutoff = FALSE)
deseq_dada2_oral_res <- cbind(deseq_dada2_oral_res, 
                              as(tax_table(ps_dada2_oral)[rownames(deseq_dada2_oral_res), ], "matrix"))
EnhancedVolcano(deseq_dada2_oral_res,
  lab = paste0(deseq_dada2_oral_res$Genus, " ", deseq_dada2_oral_res$Species),
  x = 'log2FoldChange',
  y = 'pvalue')

deseq_dada2_nasal = phyloseq_to_deseq2(ps_dada2_nasal, ~ status)
deseq_dada2_nasal = estimateSizeFactors(deseq_dada2_nasal, type = 'poscounts')
deseq_dada2_nasal = DESeq(deseq_dada2_nasal, test="Wald", fitType="parametric")

deseq_dada2_nasal_res = results(deseq_dada2_nasal, cooksCutoff = FALSE)
deseq_dada2_nasal_res <- cbind(deseq_dada2_nasal_res, 
                              as(tax_table(ps_dada2_nasal)[rownames(deseq_dada2_nasal_res), ], "matrix"))
EnhancedVolcano(deseq_dada2_nasal_res,
  lab = paste0(deseq_dada2_nasal_res$Genus, " ", deseq_dada2_nasal_res$Species),
  x = 'log2FoldChange',
  y = 'pvalue')

```

