---
title: "botero_analysis"
author: "Sean Lu"
date: "`r Sys.Date()`"
output: html_document
---
# Cleaning up metadata
```{r metadata}
metadata <- read.table("/projectsp/f_wj183_1/work/Sean/botero_16S/scripts/metadata.tsv", sep = "\t", 
                       header = TRUE)
metadata$status <- c(rep("TB", times = 18), rep("Control", times = 24), rep("TB", times = 18))
```


# Loading in metascope id tables and generating counts tables
```{r Counts table}
library(purrr)
library(dplyr)

botero_paths <- list.files("/projects/f_wj183_1/work/Sean/botero_16S/results", 
                                      full.names = TRUE, 
                                      recursive = TRUE, 
                                      pattern = ".metascope_id.csv")

process_metascope_id <- function(meta_id_path) {
  sample_name <- sub('\\..*$', '', basename(meta_id_path))
  df <- read.csv(meta_id_path)
  df <- df |>
    dplyr::select(TaxonomyID, read_count) |>
    dplyr::rename(!!sample_name := read_count)
  return(df)
}

counts_table <- purrr::reduce(lapply(botero_paths, process_metascope_id), dplyr::full_join, by = 'TaxonomyID')
counts_table[is.na(counts_table)] <- 0
```

# Generating Taxonomy Tables 
```{r Taxonomy Table}
accession_path <- "/projects/f_wj183_1/reflib/2024_accession_taxa/accessionTaxa.sql"

tax_table <- taxonomizr::getTaxonomy(counts_table$TaxonomyID, accession_path) |> as.data.frame()
tax_table <- tax_table |>
  tibble::rownames_to_column(var = "TaxonomyID") |>
  dplyr::mutate(TaxonomyID = as.numeric(TaxonomyID)) |>
  dplyr::relocate(TaxonomyID)

tax_table 
```

# Saving results
```{r eval=FALSE}
write.csv(counts_table, "/projects/f_wj183_1/work/Sean/botero_16S/analysis/counts_table.csv", 
          row.names = FALSE)
write.csv(tax_table, "/projects/f_wj183_1/work/Sean/botero_16S/analysis/tax_table.csv",
          row.names = FALSE)
write.csv(metadata, "/projects/f_wj183_1/work/Sean/botero_16S/analysis/metadata.csv",
          row.names = FALSE)
```

# Generating phyloseq objects
```{r}
library("phyloseq")
library("ggplot2")      # graphics      # necessary to import the data from Excel file
library("dplyr")        # filter and reformat data frames
library("tibble")  

otu_mat <- counts_table %>%
  tibble::column_to_rownames("TaxonomyID") 
tax_mat <- tax_table %>% 
  tibble::column_to_rownames("TaxonomyID")
samples_df <- metadata %>% 
  tibble::column_to_rownames("Run_ID") 

otu_mat <- as.matrix(otu_mat)
tax_mat <- as.matrix(tax_mat)

OTU = otu_table(otu_mat, taxa_are_rows = TRUE)
TAX = tax_table(tax_mat)
samples = sample_data(samples_df)

ps <- phyloseq(OTU, TAX, samples)
ps_oral <- subset_samples(ps, Sample_type == "Oropharynx")
total = median(sample_sums(ps_oral))
standf = function(x, t=total) round(t * (x / sum(x)))
ps_norm = transform_sample_counts(ps_oral, standf)


```

# Plotting Relative Abundance
```{r}
library(ggplot2)
p1 <- plot_bar(ps_norm, fill = "species") + 
  geom_bar(aes(color=species, fill=species), stat="identity", position="stack") +
  theme(legend.position="none") + 
  facet_wrap(~status, scales='free')
p1

```

# PLotting alpha diversity 
```{r}
library(ggpubr)
p2_1 <- plot_richness(ps_norm, measures=c("Observed", "Shannon", "Simpson", "InvSimpson"), x = "status") + 
  stat_compare_means(label = "p.signif", label.x = 1.5, comparisons = list(c("Control", "TB"))) + 
  geom_boxplot()
p2_1

p2_2 <- plot_richness(ps_norm, measures=c("Chao1", "Shannon"), x = "status")
p2_3 <- plot_richness(ps_norm, measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"), x = "status")
```
There is no change in observed alpha diversity, but there is significantly lower alpha diversity in the TB groups using Shannon, Simpson, and Inv Simpson metrics

# Plotting Ordination
```{r}
ps_norm.ord <- ordinate(ps_norm, "PCoA", "bray")
plot_ordination(ps_norm, ps_norm.ord, type="samples", color="status", 
                title="Samples") + geom_point(size=3)

```
PCoA analysis groups most TB samples closely together

# DESeq2 Analysis
```{r}
library("DESeq2")
deseq_oral = phyloseq_to_deseq2(ps_oral, ~ status)
deseq_oral = DESeq(deseq_oral, test="Wald", fitType="parametric")

res = results(deseq_oral, cooksCutoff = FALSE)
alpha = 0.1
sigtab = res[which(res$padj < alpha), ]

sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps_oral)[rownames(sigtab), ], "matrix"))
head(sigtab)
dim(sigtab)

library("ggplot2")
# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$phylum = factor(as.character(sigtab$phylum), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$species, function(x) max(x))
x = sort(x, TRUE)
sigtab$species = factor(as.character(sigtab$species), levels=names(x))
ggplot(sigtab, aes(x=species, y=log2FoldChange, color=phylum)) + geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))

```

```{r}
library(dada2)
botero_fastqs <- list.files("/projects/f_wj183_1/data/2014_botero_TB_microbiome/new_fastq", 
                                      full.names = TRUE, 
                                      recursive = TRUE, 
                                      pattern = ".fastq")

fnFs <- sort(botero_fastqs)
sample.names <- sapply(strsplit(basename(fnFs), ".fastq"), `[`, 1)
plotQualityProfile(fnFs[1:2])
```
Filter and trim Reads
```{r, eval=FALSE}
filtFs <- file.path("/projects/f_wj183_1/work/Sean/botero_16S/analysis/dada2", "filtered", 
                    paste0(sample.names, "filt.fastq.gz"))
names(filtFs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, truncLen=240,
              maxN=0, maxEE=2, truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE)
```
Filtered and trimmed reads results
```{r}
head(out)
```

```{r}
errF <- learnErrors(filtFs, multithread=TRUE)
plotErrors(errF, nominalQ=TRUE)
```


```{r}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)

seqtab <- makeSequenceTable(dadaFs)
dim(seqtab)
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)

dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```

Tracking reads through processing
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

```{r}
taxa <- assignTaxonomy(seqtab.nochim, 
                       "/projects/f_wj183_1/reflib/2024_Dada2_silva/silva_nr99_v138.2_toSpecies_trainset.fa.gz", 
                       multithread=TRUE, tryRC=TRUE)

ps_dada2 <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(samples_df), 
               tax_table(taxa))
ps_oral_dada2 <- subset_samples(ps_dada2, Sample_type == "Oropharynx")
total = median(sample_sums(ps_oral_dada2))
standf = function(x, t=total) round(t * (x / sum(x)))
ps_norm_dada2 = transform_sample_counts(ps_oral_dada2, standf)
```

# Plotting Relative Abundance
```{r}
library(ggplot2)
p1 <- plot_bar(ps_norm_dada2, fill ="Species") + 
  geom_bar(aes(color=Species, fill=Species), stat="identity", position="stack") +
  theme(legend.position="none") + 
  facet_wrap(~status, scales='free')
p1

```
# PLotting alpha diversity 
```{r}
library(ggpubr)
p2_1 <- plot_richness(ps_norm_dada2, measures=c("Observed", "Shannon", "Simpson", "InvSimpson"), x = "status") + 
  stat_compare_means(label = "p.signif", label.x = 1.5, comparisons = list(c("Control", "TB"))) + 
  geom_boxplot()
p2_1

p2_3 <- plot_richness(ps_norm_dada2, measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"), x = "status") + 
  stat_compare_means(label = "p.signif", label.x = 1.5, comparisons = list(c("Control", "TB"))) + 
  geom_boxplot()
p2_3

p2_4 <- plot_richness(ps_norm, measures=c("Observed", "Chao1", "ACE", "Shannon", "Simpson", "InvSimpson", "Fisher"), x = "status") + 
  stat_compare_means(label = "p.signif", label.x = 1.5, comparisons = list(c("Control", "TB"))) + 
  geom_boxplot()
p2_4
```


```{r}
saveRDS(ps_dada2, file = "/projects/f_wj183_1/work/Sean/botero_16S/analysis/ps_dada2.rds")
saveRDS(ps, file = "/projects/f_wj183_1/work/Sean/botero_16S/analysis/ps_metascope_priors.rds")
```

```{r}

ps_norm_dada2.ord <- ordinate(ps_norm_dada2, "PCoA", "bray")
plot_ordination(ps_norm_dada2, ps_norm_dada2.ord, type="samples", color="status", 
                title="Samples") + geom_point(size=3)
```

```{r}
library("DESeq2")
deseq_oral_dada2 = phyloseq_to_deseq2(ps_oral_dada2, ~ status)
counts_mat <- counts(deseq_oral_dada2)
counts_mat_plus1 <- counts_mat + 1
counts(deseq_oral_dada2) <- matrix(as.integer(counts_mat_plus1),
                      nrow = nrow(counts_mat_plus1),
                      ncol = ncol(counts_mat_plus1),
                      dimnames = dimnames(counts_mat_plus1))

deseq_oral_dada2 = DESeq(deseq_oral_dada2, test="Wald", fitType="parametric")

res = results(deseq_oral_dada2, cooksCutoff = FALSE)
alpha = 0.1
sigtab = res[which(res$padj < alpha), ]

sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps_oral_dada2)[rownames(sigtab), ], "matrix"))
head(sigtab)
dim(sigtab)

library("ggplot2")
# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$Species, function(x) max(x))
x = sort(x, TRUE)
sigtab$Species = factor(as.character(sigtab$Species), levels=names(x))
ggplot(sigtab, aes(x=Species, y=log2FoldChange, color=Phylum)) + geom_point(size=6) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
```